content/blog/2021-07-06-hyperparam-tuning.md

```diff
@@ -0,0 +1,280 @@
+
+```dvc
+$ python -m venv .venv
+```
+
+After you've cloned the repo, install all of the dependencies with this command.
+
+`pip install -r requirements.txt`
+
+You should be able to open your terminal and run an experiment with the
+following command.
+
+`dvc exp run`
+
+This will trigger the training process to run and it will record the precision
+of your model. You can check out the results of your experiment with the
+following command.
+
+`dvc exp show --no-timestamp --include-params train.n_est,train.min_split`
+
+_We're adding a few options here to make the table view clearer. We aren't
+showing timestamps and we're only looking at two hyperparameter values. You can
+run `dvc exp show` without the options to see the entire table._
+
+This will produce a table similar to this.
+
+```
+┏━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
+┃ Experiment              ┃ avg_prec ┃ roc_auc ┃ train.n_est ┃ train.min_split ┃
+┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
+│ workspace               │ 0.51682  │ 0.93819 │ 175         │ 64              │
+│ master                  │ 0.56447  │ 0.94713 │ 100         │ 64              │
+│ └── a1e8716 [exp-09074] │ 0.57333  │ 0.94801 │ 100         │ 32              │
+└─────────────────────────┴──────────┴─────────┴─────────────┴─────────────────┘
+```
+
+### Start tuning with grid search
+
+Now that you've seen how to run an experiment, we're going to write a small
+script to automate grid search for us using DVC. We'll do this by creating
+queues.
+
+A queue is how DVC allows us to create experiments that won't be run until
+later. That way we can cycle through multiple hyperparameters quickly instead of
+manually updating a config file with new hyperparameter values for each
+experiment run. The command syntax for creating queues looks like this:
+
+```dvc
+dvc exp run --queue -set-param train.min_split=8
+```
+
+In the example queue above, we're updating the `train.min_split` value that's
+inside of the `params.yaml` file. This file holds all of the hyperparameter
+values and is where DVC looks to determine if any values have changed. With the
+command above, we're automatically updating that value in the `params.yaml`
+using a queued experiment.
+
+Now we can make the script. You can add a new file to the `src` directory called
+`tuning.py`. Inside of the file, add the following code.
+
+```python
+import itertools
+import subprocess
+import random
+
+# Automated grid search experiments
+param_grid = {
+    "n_est_values": [250, 300, 350, 400, 450, 500],
+    "min_split_values": [8, 16, 32, 64, 128, 256]
+}
+param_combos = [dict(zip(param_grid.keys(), v)) for v in itertools.product(*param_grid.values())]
+
+for params in param_combos:
+  subprocess.run(["dvc", "exp", "run", "--queue", "-S", f"train.n_est={params['n_est_values']}", "-S", f"train.min_split={params['min_split_values']}"])
```

Comment:
Another suggestion by me that looks too complicated now 😄 . What about this?

```
# Set hyperparameter values to try.
n_est_values = [250, 300, 350, 400, 450, 500]
min_split_values = [8, 16, 32, 64, 128, 256]

# Iterate over all combinations of hyperparameter values.
for n_est, min_split in itertools.product(n_est_values, min_split_values):
    # Execute "dvc exp run --queue -S train.n_est=<n_est> -S train.min_split=<min_split>".
    subprocess.run(["dvc", "exp", "run", "--queue", "-S", f"train.n_est={n_est}", "-S", f"train.min_split={min_split}"])
```

Let me know what you think is readable. If it's hard for you to read, it's probably hard for other people. I don't like using `itertools.product` here, but I'm not sure there's another succinct way to do it. There's https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.ParameterGrid.html, but I'm not sure it makes sense to have a scikit-learn dependency here.

Also, do you think we need to introduce the short `-S` option?

Reply:
I think this looks more readable. I can definitely follow it a little easier!

I decided to just use the long version `--set-param`. I think that'll keep it clearer for people just reading along.